core/preconditioner/gauss_seidel.cpp

```diff
@@ -0,0 +1,235 @@
+{
+    if (&other != this) {
+        EnableLinOp<GaussSeidel>::operator=(other);
+        auto exec = this->get_executor();
+        solver_comp_ = other.solver_comp_;
+        parameters_ = other.parameters_;
+        if (solver_comp_ && other.solver_comp_->get_executor() != exec) {
+            solver_comp_ = gko::clone(exec, solver_comp_);
+        }
+    }
+    return *this;
+}
+
+
+template <typename ValueType, typename IndexType>
+GaussSeidel<ValueType, IndexType>& GaussSeidel<ValueType, IndexType>::operator=(
+    GaussSeidel&& other)
+{
+    if (&other != this) {
+        EnableLinOp<GaussSeidel>::operator=(std::move(other));
+        auto exec = this->get_executor();
+        solver_comp_ = std::move(other.solver_comp_);
+        parameters_ = std::exchange(other.parameters_, parameters_type{});
+        if (solver_comp_ && other.solver_comp_->get_executor() != exec) {
+            solver_comp_ = gko::clone(exec, solver_comp_);
+        }
+    }
+    return *this;
+}
+
+
+template <typename ValueType, typename IndexType>
+GaussSeidel<ValueType, IndexType>::GaussSeidel(const GaussSeidel& other)
+    : GaussSeidel{other.get_executor()}
+{
+    *this = other;
+}
+
+
+template <typename ValueType, typename IndexType>
+GaussSeidel<ValueType, IndexType>::GaussSeidel(GaussSeidel&& other)
+    : GaussSeidel{other.get_executor()}
+{
+    *this = std::move(other);
+}
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> GaussSeidel<ValueType, IndexType>::transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> GaussSeidel<ValueType, IndexType>::conj_transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+template <typename ValueType, typename IndexType>
+void GaussSeidel<ValueType, IndexType>::generate(
+    std::shared_ptr<const LinOp> system_matrix, bool skip_sorting)
+{
+    using Csr = matrix::Csr<ValueType, IndexType>;
+    const auto exec = this->get_executor();
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+
+    auto csr_matrix =
+        convert_to_with_sorting<Csr>(exec, system_matrix, skip_sorting);
+
+    const auto mat_size = csr_matrix->get_size();
+    const auto num_rows = mat_size[0];
+    auto mat_strategy = csr_matrix->get_strategy();
+    if (!symmetric_preconditioner_) {
+        // init row pointers
+        array<IndexType> l_row_ptrs{exec, num_rows + 1};
+        exec->run(gauss_seidel::make_initialize_row_ptrs_l(
+            csr_matrix.get(), l_row_ptrs.get_data()));
+
+        // Get nnz from device memory
+        auto l_nnz = static_cast<size_type>(
+            exec->copy_val_to_host(l_row_ptrs.get_data() + num_rows));
+
+        // Init arrays
+        array<IndexType> l_col_idxs{exec, l_nnz};
+        array<ValueType> l_vals{exec, l_nnz};
+        std::shared_ptr<Csr> l_factor = Csr::create(
+            exec, mat_size, std::move(l_vals), std::move(l_col_idxs),
+            std::move(l_row_ptrs), mat_strategy);
```

Comment:
```suggestion
        std::shared_ptr<Csr> l_factor = Csr::create(
            exec, mat_size, array<IndexType>{exec, l_nnz},  
            array<IndexType>{exec, l_nnz}, std::move(l_row_ptrs), 
            mat_strategy);
```
I don't think, the arrays are needed outside the matrix. Same for the symmetric case

Reply:
Ok, Should i change the code for ic/ilu as well? I adjusted the code from core/ic.cpp / core/ilu.cpp and there the arrays also don't seem to be needed aside from the init_l_u call...

Reply:
Sure, go ahead

Reply:
But do that in a separate commit